src/blas/impl/KokkosBlas3_gemm_impl.hpp

```diff
@@ -0,0 +1,506 @@
+      }
+      C(i,B_j) += C_ij0;
+      C(i,B_j+blockB1/4) += C_ij1;
+      C(i,B_j+2*blockB1/4) += C_ij2;
+      C(i,B_j+3*blockB1/4) += C_ij3;
+#if defined(__CUDA_ARCH__) || !defined(KOKKOS_ENABLE_OPENMP)
+    });
+#else
+    }
+#endif
+  });
+}
+
+template<int TransposeA, int TransposeB>
+struct impl_gemm_label;
+
+template<>
+struct impl_gemm_label<0,0> {
+  static constexpr const char* label = "KokkosBlas::gemm[NN]";
+};
+template<>
+struct impl_gemm_label<0,1> {
+  static constexpr const char* label = "KokkosBlas::gemm[NT]";
+};
+template<>
+struct impl_gemm_label<0,2> {
+  static constexpr const char* label = "KokkosBlas::gemm[NC]";
+};
+
+template<>
+struct impl_gemm_label<1,0> {
+  static constexpr const char* label = "KokkosBlas::gemm[TN]";
+};
+template<>
+struct impl_gemm_label<1,1> {
+  static constexpr const char* label = "KokkosBlas::gemm[TT]";
+};
+template<>
+struct impl_gemm_label<1,2> {
+  static constexpr const char* label = "KokkosBlas::gemm[TC]";
+};
+
+template<>
+struct impl_gemm_label<2,0> {
+  static constexpr const char* label = "KokkosBlas::gemm[CN]";
+};
+template<>
+struct impl_gemm_label<2,1> {
+  static constexpr const char* label = "KokkosBlas::gemm[CT]";
+};
+template<>
+struct impl_gemm_label<2,2> {
+  static constexpr const char* label = "KokkosBlas::gemm[CC]";
+};
+
+
+template<class ExecSpace, class ViewTypeA, class ViewTypeB, class ViewTypeC,
+          int blockA0, int blockA1, int blockB1, int TransposeA, int TransposeB>
+struct GEMMImpl {
+  ViewTypeA A;
+  ViewTypeB B;
+  ViewTypeC C;
+  typedef typename ViewTypeA::non_const_value_type ScalarA;
+  typedef typename ViewTypeB::non_const_value_type ScalarB;
+  typedef typename ViewTypeC::non_const_value_type ScalarC;
+
+  const int num_blocks_0;
+  const int num_blocks_1;
+  int scratch_level;
+
+  ScalarC alpha, beta;
+  typedef Kokkos::View<ScalarA[blockA0][blockA1],Kokkos::LayoutLeft,typename ExecSpace::scratch_memory_space>
+    ViewTypeAScratch;
+  typedef Kokkos::View<ScalarB[blockA1][blockB1],Kokkos::LayoutRight,typename ExecSpace::scratch_memory_space>
+    ViewTypeBScratch;
+  typedef Kokkos::View<ScalarC[blockA0][blockB1],Kokkos::LayoutRight,typename ExecSpace::scratch_memory_space>
+    ViewTypeCScratch;
+
+  GEMMImpl(const ScalarA& alpha_, const ViewTypeA& A_, const ViewTypeB& B_, const ScalarC& beta_, const ViewTypeC& C_):A(A_),B(B_),C(C_),
+      num_blocks_0((C.extent(0)+blockA0-1)/blockA0),num_blocks_1((C.extent(1)+blockB1-1)/blockB1) {
+    scratch_level = 0;
+    alpha = alpha_;
+    beta = beta_;
+  }
+
+  void run(int team_size, int vector_length, int scr_level) {
+    scratch_level = scr_level;
+    int scratch_memory_size =
+      ViewTypeAScratch::required_allocation_size() +
+      ViewTypeBScratch::required_allocation_size() +
+      ViewTypeCScratch::required_allocation_size();
+
+    Kokkos::TeamPolicy<ExecSpace,Kokkos::LaunchBounds<384,2>> policy(num_blocks_0*num_blocks_1,team_size,vector_length);
+
```

Comment:
How can you handle different shapes of gemm ? For instance, this parallel run would not work well when C matrix is small as parallelism comes from breaking C into pieces (I call this FixC algorithm). Maybe later other algorithms (FixA, FixB) would be beneficial but I also think that we are not saving the world. So, current impl should be okay.

Reply:
Yeah if there are two few C blocks this is not gonna be super great. Right now its broken up in chunks of 24x64. One thing we could do as a stop-gap measure is to instantiate also a 8x16 variant or so. That said on say Haswell dual socket we should be good as long as C is at least something like 200x200. On P100 about 300x300.